mpi/TOASTFiller.py

```diff
@@ -0,0 +1,98 @@
+from toast.tod import TODCache
+from toast import qarray as qa
+from spt3g import core, calibration, dfmux, coordinateutils
+import numpy
+from .MPIAccumulator import MPIAccumulator
+import pickle
+
+class TOASTFiller(object):
+    '''
+    Fill a TOAST TOD (stored in the toastobses member after completion) from
+    a frame stream. Data will be distributed across the given MPI communicator
+    in the way that TOAST would like.
+    '''
+    def __init__(self, mpicomm, boresight='OnlineRaDecRotation', timestreams='RawTimestreams_I', wiring='WiringMap'):
+        self.boresight = boresight
+        self.timestreams = timestreams
+        self.wiring = wiring
+        self.mpicomm = mpicomm
+        self.mpia = MPIAccumulator(mpicomm, self.extractor, lambda m: m[0]['start'], dataframes=[core.G3FrameType.Scan])
+
+    def extractor(self, frame):
+        if frame.type == core.G3FrameType.Wiring:
+            self.keys = frame[self.wiring].keys()
+            self.invkeys = {k: i for i,k in enumerate(self.keys)}
+
+        if frame.type == core.G3FrameType.Calibration:
+            #self.detquat = {k: qa.from_angles(v.y_offset/core.G3Units.rad, v.x_offset/core.G3Units.rad, 0) for k,v in frame['BolometerProperties'].items()}
+            self.detquat = None
+
+        if frame.type == core.G3FrameType.Scan:
+            timestreams = frame[self.timestreams]
+            bs = numpy.asarray(frame[self.boresight])
+            bs = bs[:,[1,2,3,0]] # Swizzle for TOAST
+            scan = {'dets': self.keys, 'boresight': bs, 'detquat': self.detquat, 'obsid': '%s-%d' % (frame['SourceName'], frame['ObservationID']), 'start': timestreams.start, 'stop': timestreams.stop, 'nsamples': len(timestreams.values()[0])}
+            return scan['obsid'], scan, timestreams
+
+
+    def __call__(self, frame):
+        rv = self.mpia(frame)
+        if frame.type != core.G3FrameType.EndProcessing:
+            return rv
+
+        # Build TOAST TOD from our full data set, now that we have it
+        toastobses = []
+        for obs,scans in self.mpia.fullobs.items():
+            nsamptot = sum([i[0]['nsamples'] for i in scans])
```

Comment:
I agree with the general paradigm of the TOASTFiller class, but I think it might be hard to make a general class for this, since every experiment has different toast.TOD derived classes and would need a custom extraction class.

Imagine that an experiment has a bunch of observations, each with lots of frames.  Some external function from the experiment determines how to divide up those frames among some MPI processes.  Now every process runs a G3Pipeline on its frames.  Next:

1.  The experiment defines an "extractor" function that enforces how frames should be unpacked according to naming conventions, etc.

2.  The experiment defines a toast.TOD derived class with some constructor kwargs

3.  A generic "TOASTFiller" class could then take as constructor arguments the extractor class instance, the TOD class type, and the common kwargs.  It would then use MPIAccumulator on the toast group communicator with the specified extractor function and for each observation it would instantiate the experiment TOD class and fill the local cache with the unpacked frame data.

Alternatively, if more specialization was needed, the experiment could just have their own custom TOASTFiller class that did extra things.  I'm not sure if that is needed, but we could try to implement something like this for an experiment and see how it works.

Reply:
This is a tricky one. I think I like your idea, and will take a shot at implementing it. Since the TOASTFiller is experimental anyway, I might merge the rest of this and then poke at it separately, if you don't mind.